### ISSUE TITLE
Metropolis fails to converge on simple model (PyMC 3)

### ISSUE THREAD

**Issue by fonnesbeck** [pymc-devs/pymc3#358]
I have attempted to replicate a very simple partial pooling model from Gelman and Hill in PyMC 3, but find that the `Metropolis` step method fails to converge even in a large number of iterations. NUTS seems to handle it fine, but there appears to be an issue with Metropolis. Perhaps related to adaptation?

[Here is a link to a zip file](https://dl.dropboxusercontent.com/u/233041/multilevel.zip) containing the model and data.

**Comment by fonnesbeck** [comment]
I have verified that the equivalent model in PyMC 2 converges to reasonable values under Metropolis (using Gelman and Hill's result as the gold standard), so something appears to be up with the PyMC 3 implementation.

Edited: figured out the NUTS issue

**Comment by jsalvatier** [comment]
Any update?

**Comment by fonnesbeck** [comment]
No, still a mystery.

**Comment by bjedwards** [comment]
Does the model still work? I get a `AsTensorError` on line 49, when trying to call `y_hat = a[county]`. I am running theano off the github-dev.

**Comment by aflaxman** [comment]
@bjedwards it was giving me this error also, and adding `county = county.astype(int)` before setting up the model made the error go away.  I can confirm that Metropolis is not converging.

![image](https://f.cloud.github.com/assets/51236/1699480/8ef4770c-5fa0-11e3-9db9-c49e6d988372.png)

@fonnesbeck do you have a PyMC 2 version for comparison?

**Comment by aflaxman** [comment]
Given long enough, it seems to work for me.  I think I know what the issue is, but I'll decide for sure when I see what PyMC2 is doing for @fonnesbeck ---

![image](https://f.cloud.github.com/assets/51236/1720032/dd70e25a-61f6-11e3-83f1-14fa1f789c8a.png)

**Comment by fonnesbeck** [comment]
Sorry. Will have a go at this tonight.

**Comment by aflaxman** [comment]
No worries.  I was suspicious that the PyMC2 defaults would actually correspond more to a CompoundStep of Metropoli in PyMC3.  However, that doesn't converge any faster:

```
with partial_pooling:
    steps = []
    steps.append(pm.Metropolis(vars=[mu_a], tune_interval=1000))
    steps.append(pm.Metropolis(vars=[sigma_a], tune_interval=1000))
    steps.append(pm.Metropolis(vars=[a], tune_interval=1000))
    steps.append(pm.Metropolis(vars=[sigma_y], tune_interval=1000))

    partial_pooling_samples = pm.sample(10000, pm.CompoundStep(steps))
```

![image](https://f.cloud.github.com/assets/51236/1720191/228650ce-61fb-11e3-892a-fbe2350a3b5b.png)

**Comment by fonnesbeck** [comment]
The following implementation in PyMC2 seems to work, using `AdaptiveMetropolis` for sampling the random intercepts:

```
def partial_pooling():

    # Priors
    mu_a = Normal('mu_a', mu=0., tau=0.0001, value=0)
    sigma_a = Uniform('sigma_a', lower=0, upper=100, value=1.)
    tau_a = sigma_a**-2

    # Random intercepts
    a = Normal('a', mu=mu_a, tau=tau_a, value=np.zeros(len(set(county))))

    # Model error
    sigma_y = Uniform('sigma_y', lower=0, upper=100)
    tau_y = sigma_y**-2

    # Expected value
    y_hat = Lambda('y_hat', lambda a=a: a[county])

    # Data likelihood
    y_like = Normal('y_like', mu=y_hat, tau=tau_y, value=log_radon, observed=True)

    return locals()
```

Going to go back and try again in PyMC3.

**Comment by twiecki** [comment]
This works fine for me on this branch: https://github.com/pymc-devs/pymc/pull/587 Please reopen if problem persists.

**Comment by lwahedi** [comment]
I ran into some trouble with this and wanted to post a note for later readers who are also having trouble. I made some toy data for a pretty simple partial pooling model based on the one discussed here, NUTS converges on the true values in not that many iterations (a few hundred), but very slowly. (I'm having some serious stalling issue on my actual data with NUTS.) For the first ten thousand or so iterations Metropolis settles over 0. At around 40k it's finally centered over the true value. At 100k, it's still a flat topped curve with a peak that isn't over the true value, and credible intervals that still cross 0. And this is with a pretty large effect size in the toy data. At 150k it starts to tighten, but still isn't really there yet for all the parameters. But 150k is still much faster than a couple hundred on NUTS, 
TLDR: Metropolis needs a few hundred thousand iterations to converge on this model.

**Comment by twiecki** [comment]
Thanks @lwahedi. Have you used ADVI initialization? I would also set tune=100 for NUTS. You can also investigate the sampler stats to identify difficult regions: http://pymc-devs.github.io/pymc3/notebooks/Diagnosing_biased_Inference_with_Divergences.html

**Comment by fonnesbeck** [comment]
@lwahedi  If you just run `trace = sample(2000)` (for example), PyMC will automatically select NUTS and initialize it with ADVI automatically.

**Comment by lwahedi** [comment]
Thanks for the tips. ADVI does make this particular model run much faster. It still stalls when I add a probit transformation on the outcome, but I've now learned that Metropolis crashes the kernel at somewhere between 30k-50k iterations on the extended model too, so that wasn't the solution I'd hoped it would be. Either way, off topic for this issue. [I have a stackexchange question with more detail about it here. ](http://stackoverflow.com/questions/42779123/convergence-issues-on-hierarchical-probit-model-with-nuts-and-metropolis). It may be a specification issue--not sure what's going on.
